### PR TITLE
EDGE-52: update call center sample app to use new signalling

### PR DESCRIPTION
**Pull request by cejohnson-bw**

### CODE REVIEW

File: src/server.ts [hunk 1]
```diff
@@ -3,33 +3,56 @@ import bodyParser from "body-parser";
...
- * Initialize the Bandwidth WebRTC SDK
- */
-bandwidthRtc.getClientCredentials({
-  authorizationEndpoint: <string>process.env.AUTHORIZATION_ENDPOINT, // TODO: Remove this URL after it is built into the SDK
-  clientId: <string>process.env.CLIENT_ID,
-  clientSecret: <string>process.env.CLIENT_SECRET
-});
+const agents: Map<string, Agent> = new Map();
+const conferences: Map<string, string> = new Map();
+const participants: Map<string, string> = new Map();
```
Comment (cejohnson-bw) [inline src/server.ts]: I forgot I did this.... it can probably be cleaned up a lot.

File: src/server.ts [hunk 2]
```diff
@@ -3,33 +3,56 @@ import bodyParser from "body-parser";
...
+  .connect({
+    accountId: accountId,
+    username: username,
+    password: password
+  })
+  .then(() => {
+    console.log("bandwidth rtc websocket connected");
+  })
+  .catch(error => {
+    console.log("Error connecting to bandwidthRtc", error.message);
```
Comment (nguyer) [inline src/server.ts]: Consider rewording "connecting to bandwidthRtc"

Maybe "Error connecting bandwidth rtc websocket"

File: src/server.ts [hunk 3]
```diff
@@ -42,7 +65,16 @@ const callQueue: string[] = [];
  * If we wanted to support multiple agents taking calls from this queue, we would want this to be a map
  * to track which call each agent is currently on.
  */
-let activeCallId: string | undefined;
+let activeCalls: Map<string, string> = new Map();
+
+const availableAgent = () => {
+  for (const agent of agents.values()) {
```
Comment (nguyer) [inline src/server.ts]: Do you have to use `.values()` here or will it let you do a `for...of` over a map natively? I'm not sure.

File: src/server.ts [hunk 4]
```diff
@@ -93,53 +131,73 @@ const connectCall = async (callId: string) => {
...
+      if (agent.conferenceId) {
+        await bandwidthRtc.endConference(agent.conferenceId);
+      }
+      agent.online = false;
+      agent.conferenceId = undefined;
+      agent.participantId = undefined;
+      agent.streamId = undefined;
+      agent.activeCallId = undefined;
+    }
+    res.status(200).send(resp);
```
Comment (nguyer) [inline src/server.ts]: There is a lot going on here for a route handler. Maybe we should break this out into two separate functions for going online or offline?

File: src/server.ts [hunk 5]
```diff
@@ -167,22 +225,25 @@ app.post("/callback/incoming", (req, res) => {
   console.log(`New incoming call ${callId}`);
   let bxml = "";
 
-  if (!activeCallId && agentOnline) {
+  let agent = availableAgent();
+  if (agent) {
     // An agent is online and waiting, so immediately connect this call
-    activeCallId = callId;
+    // activeCallId = callId;
```
Comment (nguyer) [inline src/server.ts]: dead code?

File: src/server.ts [hunk 6]
```diff
@@ -212,19 +273,29 @@ app.post("/callback/hold", (req, res) => {
...
+    if (agent.activeCallId === callId) {
+      let conferenceId = agent.conferenceId;
+      if (conferenceId) {
+        console.log(`Connecting call ${callId} to agent ${agent.id}`);
+        let participantId = await bandwidthRtc.createParticipant(conferenceId);
+        participants.set(participantId, callId);
+        const bxml = `<?xml version="1.0" encoding="UTF-8" ?>
+        <Response>
+            <Transfer transferCallerId="+1${conferenceId}${participantId}">
+                <PhoneNumber>+19192892727</PhoneNumber>
```
Comment (nguyer) [inline src/server.ts]: This is fine for prod, but won't work in edge

File: src/server.ts [hunk 7]
```diff
@@ -3,33 +3,56 @@ import bodyParser from "body-parser";
...
-app.use(bodyParser.json());
+interface Agent {
+  id: string;
+  online: boolean;
+  conferenceId?: string;
+  participantId?: string;
+  streamId?: string;
+  activeCallId?: string;
+  username?: string;
+}
```
Comment (michaelcaldwell) [inline src/server.ts]: like

File: src/server.ts (hunk 1 above)
Comment (michaelcaldwell) [reply src/server.ts]: I dunno.  If they're all used for a different key lookup, a-ok

File: src/server.ts (hunk 2 above)
Comment (michaelcaldwell) [reply src/server.ts]: suggest:
```
"Bandwidth WebRTC socket connected"
```
and
```
"Error connecting to Bandwidth WebRTC"
```

Comment (michaelcaldwell) [reply src/server.ts]: thinka now we shouldn't use the word "rtc" by itself anywhere

File: src/server.ts [hunk 8]
```diff
@@ -93,53 +131,73 @@ const connectCall = async (callId: string) => {
...
 
 /**
  * The browser will call this endpoint to update their status (online/offline)
  */
 app.post("/setStatus", async (req: any, res) => {
-  if (req.body.online) {
-    console.log(`The agent is coming online`);
-    agentOnline = true;
-    connectNextCall();
+  let agent = agents.get(req.body.token);
```
Comment (michaelcaldwell) [inline src/server.ts]: so, a token is an agent?

File: src/server.ts [hunk 9]
```diff
@@ -93,53 +131,73 @@ const connectCall = async (callId: string) => {
...
+    res.status(200).send({token: token});
+  } catch (e) {
+    res.status(400).send(e);
+  }
 });
 
 /**
  * The browser will call this endpoint to update their status (online/offline)
  */
 app.post("/setStatus", async (req: any, res) => {
```
Comment (michaelcaldwell) [inline src/server.ts]: yeah, to @nguyer 's point, not restful or very good, but don't want to break any existing callers of this service today, ,that's out of scope.  But if we can do it now, or later, I suggest:
```
app.post("/online")
```
and
```
app.post("/offline")
```

Comment (michaelcaldwell) [reply src/server.ts]: but the most REST-ful interface, but less intuitive IMO, would be:
```
app.post("/ready")
```
and 
```
app.delete("/ready")
```

File: src/server.ts (hunk 4 above)
Comment (michaelcaldwell) [reply src/server.ts]: without breaking the current http api, I think he has no choice but to implement both cases in this handler, though it could be split into two subroutines, and you could
```
if (!agent) {
  res.status(404).send({});
}
to shortcut that case at the top and remove a nesting level for the rest of the func

File: src/server.ts [hunk 10]
```diff
@@ -93,53 +131,73 @@ const connectCall = async (callId: string) => {
...
+    let activeCallId = agent.activeCallId;
     console.log(`Hanging up ${activeCallId}`);
     let response;
     try {
       response = await axios.post(
         `https://voice.bandwidth.com/api/v2/accounts/${accountId}/calls/${activeCallId}`,
         {
-          state: "completed"
+          state: "completed",
+          applicationId: voiceAppId
```
Comment (michaelcaldwell) [inline src/server.ts]: where does voiceAppId come from?

File: src/server.ts (hunk 5 above)
Comment (michaelcaldwell) [reply src/server.ts]: dead code

File: src/server.ts [hunk 11]
```diff
@@ -212,19 +273,29 @@ app.post("/callback/hold", (req, res) => {
...
+        let participantId = await bandwidthRtc.createParticipant(conferenceId);
+        participants.set(participantId, callId);
+        const bxml = `<?xml version="1.0" encoding="UTF-8" ?>
+        <Response>
+            <Transfer transferCallerId="+1${conferenceId}${participantId}">
+                <PhoneNumber>+19192892727</PhoneNumber>
+            </Transfer>
+        </Response>`;
+        res.contentType("application/xml").send(bxml);
+        return;
```
Comment (michaelcaldwell) [inline src/server.ts]: should be res.status(200).send(); ?

File: src/server.ts [hunk 12]
```diff
@@ -212,19 +273,29 @@ app.post("/callback/hold", (req, res) => {
  * This is triggered in response to the HTTP Post above
  * that sets the call's redirectUrl
  */
-app.post("/callback/connect", (req, res) => {
+app.post("/callback/connect", async (req, res) => {
```
Comment (michaelcaldwell) [inline src/server.ts]: what does adding async to an http route actually do?  why is it needed?

File: src/server.ts [hunk 13]
```diff
@@ -212,19 +273,29 @@ app.post("/callback/hold", (req, res) => {
...
-  console.log(`Connecting call ${callId}`);
-  const bxml = `<?xml version="1.0" encoding="UTF-8" ?>
-  <Response>
-      <Transfer transferCallerId="+${sessionId}${from.replace("+", "")}">
-          <PhoneNumber>${conferenceNumber}</PhoneNumber>
-      </Transfer>
-  </Response>`;
-  res.contentType("application/xml").send(bxml);
-  console.log(bxml);
+  let agent = agents.get(activeCalls.get(callId) || '');
```
Comment (michaelcaldwell) [inline src/server.ts]: ``` || null):  ``` ?

File: src/server.ts [hunk 14]
```diff
@@ -241,17 +312,62 @@ app.post("/callback/status", (req, res) => {
   const callId = req.body.callId;
   const eventType = req.body.eventType;
   if (eventType === "disconnect") {
-    console.log(`Call ${callId} disconnected`);
+    console.log(`Call ${callId} has ended`);
     res.status(200).send();
```
Comment (michaelcaldwell) [inline src/server.ts]: send the status and then due the below things?  is the order reversed?

File: src/server.ts [hunk 15]
```diff
@@ -241,17 +312,62 @@ app.post("/callback/status", (req, res) => {
...
+  let agent = agents.get(conferences.get(event.conferenceId) || '');
+  if (agent) {
+    if (event.participantId === agent.participantId) {
+      let conferenceId = event.conferenceId;
+      await bandwidthRtc.endConference(conferenceId);
+      agent.online = false;
+      agent.conferenceId = undefined;
+      agent.participantId = undefined;
+      agent.streamId = undefined;
+      agent.activeCallId = undefined;
```
Comment (michaelcaldwell) [inline src/server.ts]: setting to null is better than undefined?

File: src/server.ts (hunk 6 above)
Comment (michaelcaldwell) [reply src/server.ts]: should be ```${conferenceNumber}``` right?

**Comment by michaelcaldwell** [comment]
small thing, but this is customer-facing, so:
on npm install:
npm WARN webrtc-customer-server-sample@1.0.0 No description
npm WARN webrtc-customer-server-sample@1.0.0 No repository field.

File: src/server.ts (hunk 3 above)
Comment (cejohnson-bw) [reply src/server.ts]: I can do `for (const [_, agent] of agents) {` if that's preferable. According to this random website,

"Use for…of to iterate over the values in an iterable, like an array"
and
"Use for…in to iterate over the properties of an object (the object keys)"

https://alligator.io/js/for-of-for-in-loops/

File: src/server.ts (hunk 9 above)
Comment (cejohnson-bw) [reply src/server.ts]: Yeah, that's how it was already, I'll come back around to it in the refactor.

File: src/server.ts (hunk 8 above)
Comment (cejohnson-bw) [reply src/server.ts]: Yeah, this is all pretty vague. I'm basically using it as poor man's auth, so I'm using the token in place of a cookie or JWT, with the idea being on the way back in it's just a way to look up the agent. I went for the simplest way to do that, which is a map of agent token (which I'm calling id on the agent... idk) to an agent object in memory.

TL;DR token == agent.id == a way to look up an agent

File: src/server.ts [hunk 16]
```diff
@@ -93,53 +131,73 @@ const connectCall = async (callId: string) => {
...
+    } else {
+      console.log(`Agent ${agent.id} is going offline`);
+      if (agent.conferenceId) {
+        await bandwidthRtc.endConference(agent.conferenceId);
+      }
+      agent.online = false;
+      agent.conferenceId = undefined;
+      agent.participantId = undefined;
+      agent.streamId = undefined;
+      agent.activeCallId = undefined;
```
[the remark this answers is not in the document]
Comment (cejohnson-bw) [reply src/server.ts]: I'm not sure. I left it this way so that I could get away with using optionals in the interface (`conferenceId?: string;`). Based on the language defaulting to `undefined` for optional values, I would assume that's at least a good practice (vs `conferenceId: string | null;`), but I really don't know.

File: src/server.ts (hunk 4 above)
Comment (cejohnson-bw) [reply src/server.ts]: It's not a big deal to break the API, the FE is comparatively simple. Splitting it makes sense.

File: src/server.ts (hunk 5 above)
Comment (cejohnson-bw) [reply src/server.ts]: Dead code

File: src/server.ts (hunk 12 above)
Comment (cejohnson-bw) [reply src/server.ts]: It's needed so that I can do this a little later
```
await bandwidthRtc.createParticipant(conferenceId);
```

File: src/server.ts (hunk 13 above)
Comment (cejohnson-bw) [reply src/server.ts]: I wish. map.get() requires a non-null key. Unfortunately I think I'm going to have to do some nesting
```
let agentId = activeCalls.get(callId);
if (agentId) {
  let agent = agents.get(agentId);
  if (agent) {
    ...
```

:\

File: src/server.ts (hunk 11 above)
Comment (cejohnson-bw) [reply src/server.ts]: Using return here to break out of the function, which will otherwise return a 400 in case any of these ifs weren't true. Not very intuitive, I'll try to simplify it.

File: src/server.ts (hunk 14 above)
Comment (cejohnson-bw) [reply src/server.ts]: This is basically how it was, and I think a neat feature of express actually. This way we can quickly respond to V2 with a 200 saying got it, and then carry on doing cleanup on our end.

File: src/server.ts (hunk 9 above)
Comment (cejohnson-bw) [reply src/server.ts]: I ended up leaving the path as is for now, the entire API (3 whole endpoints) is basically RPC, so I'd want to give it a little more thought.

File: src/webrtcAgentManager.ts [hunk 17]
```diff
@@ -0,0 +1,264 @@
+import BandwidthRtc, {
```
Comment (cejohnson-bw) [inline src/webrtcAgentManager.ts]: This is sizable, but hopefully easier to follow. By all means call out any issues.

File: src/webrtcAgentManager.ts [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+   * @param username Bandwidth username
+   * @param password Bandwidth password
+   */
+  async initialize(accountId: string, username: string, password: string) {
+    await this.bandwidthRtc.connect({
+      accountId: accountId,
+      username: username,
+      password: password
+    });
+    console.log("Bandwidth WebRTC websocket connected");
```
Comment (cejohnson-bw) [inline src/webrtcAgentManager.ts]: I don't love having the event handlers inside initialize, but I don't love having them anywhere else either.

File: src/webrtcAgentManager.ts (hunk 17 above)
Comment (cejohnson-bw) [reply src/webrtcAgentManager.ts]: Having skimmed over it, I think this whole file could probably use some more comments.

File: src/callQueue.ts [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    this.connectCall = connectCall;
+
+    this.agentManager.onAgentAvailable(this.connectNextCall.bind(this));
+    this.agentManager.onCallEnded((callId: string) => {
+      this.activeCalls.delete(callId);
+    });
+  }
+
+  /**
+   * Assign an agent to the call if one is available, or add the call queue if not.
```
Comment (cejohnson-bw) [inline src/callQueue.ts]: "call to the queue", fixed, not merged.

File: src/server.ts (hunk 3 above)
Comment (michaelcaldwell) [reply src/server.ts]: what you have is more readable

File: src/callQueue.ts [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+    console.log(`Handling the next call`);
+    if (this.callQueue.length > 0) {
+      let agentId = this.agentManager.getAvailableAgent();
+      if (agentId) {
+        let activeCallId = this.callQueue.shift();
+        if (activeCallId) {
+          this.agentManager.assignCall(agentId, activeCallId);
+          this.activeCalls.set(activeCallId, agentId);
+          console.log(`${activeCallId} is the new active call for agent ${agentId}`);
+          this.connectCall(activeCallId);
```
Comment (michaelcaldwell) [inline src/callQueue.ts]: I don't see the connectCall() method declaration

File: src/webrtcAgentManager.ts (hunk 18 above)
Comment (michaelcaldwell) [reply src/webrtcAgentManager.ts]: my ignorance, but what is wrong with having them outside of initialize()?

Comment (michaelcaldwell) [reply src/webrtcAgentManager.ts]: I mean, because we wait on this.bandwidthRtc.connect(), we know that bandwidthRtc instance will be in the proper state before the below events fire, as long as the consumer of of webrtcAgentManager calls initialize() immediately after constructing the instance.  Which makes me ask, why not just have the call to .connect() inside the constructor and do away with initialize()?  Would have to pass the 3 args to the ctor of course.

File: src/webrtcAgentManager.ts [hunk 21]
```diff
@@ -0,0 +1,264 @@
...
+      await this.bandwidthRtc.endConference(conferenceId);
+    }
+  }
+
+  /**
+   * Return the first availabe agent's id, or null
+   */
+  getAvailableAgent() {
+    // TODO: grab agents more fairly if there are multiple available
+    for (const [_, agent] of this.agents) {
```
Comment (michaelcaldwell) [inline src/webrtcAgentManager.ts]: woo!

File: src/webrtcAgentManager.ts [hunk 22]
```diff
@@ -0,0 +1,264 @@
...
+  }
+
+  /**
+   * Helper function that will generate and return the correct Bandwidth BXML to transfer a caller to an agent
+   * @param agentId agent id
+   */
+  async getTransferBxml(agentId: string) {
+    const agent = this.getAgent(agentId);
+    if (this.isOnline(agent)) {
+      const participantId = await this.bandwidthRtc.createParticipant(agent.conferenceId!);
```
Comment (michaelcaldwell) [inline src/webrtcAgentManager.ts]: what does the ! at the end of conferenceId do?

File: src/webrtcAgentManager.ts [hunk 23]
```diff
@@ -0,0 +1,264 @@
...
+    return agent.activeCallId;
+  }
+
+  /**
+   * Helper function that will generate and return the correct Bandwidth BXML to transfer a caller to an agent
+   * @param agentId agent id
+   */
+  async getTransferBxml(agentId: string) {
+    const agent = this.getAgent(agentId);
+    if (this.isOnline(agent)) {
```
Comment (michaelcaldwell) [inline src/webrtcAgentManager.ts]: first, throw a typed error here (AgentNotFound) ``` if (!agent) {  ```, before checking if the agent is online

File: src/webrtcAgentManager.ts [hunk 24]
```diff
@@ -0,0 +1,264 @@
...
+  getActiveCallId(agentId: string) {
+    const agent = this.getAgent(agentId);
+    return agent.activeCallId;
+  }
+
+  /**
+   * Helper function that will generate and return the correct Bandwidth BXML to transfer a caller to an agent
+   * @param agentId agent id
+   */
+  async getTransferBxml(agentId: string) {
```
Comment (michaelcaldwell) [inline src/webrtcAgentManager.ts]: so we talked about the name of the bandwidthRtc.generateTransferXml() method, which is fine.
but shouldn't this method be named something like transferCaller(agentId)?

File: src/server.ts [hunk 25]
```diff
@@ -199,32 +180,33 @@ app.post("/callback/hold", (req, res) => {
...
-  console.log(bxml);
+  const agentId = callQueue.getAssignedAgent(callId);
+  if (agentId) {
+    const transferBxml = await agentManager.getTransferBxml(agentId);
+    console.log(`Connecting call ${callId} to agent ${agentId}`);
+    const bxml = `<?xml version="1.0" encoding="UTF-8" ?>
+    <Response>
+        ${transferBxml}
+    </Response>`;
+    res.contentType("application/xml").send(bxml);
```
Comment (michaelcaldwell) [inline src/server.ts]: suggest encapsulating this within agentManager:
```
console.log(`Connecting call ${callId} to agent ${agentId}`);
const xmlResp = await agentManager.transferCaller(callId, agentId);
res.contentType("application/xml").send(xmlResp);
```
because this xml wrapper syntax will never change, and it's a real magic string as we have it now, and server app just doesn't need to know this level of detail

File: src/callQueue.ts (hunk 20 above)
Comment (cejohnson-bw) [reply src/callQueue.ts]: Yeah, it's passed in as a callback since it's doing V2 things, which I was trying to keep out of here. Don't love the indirection though. I also in general don't love both the triangle architecture I ended up with where server instantiates an AgentManager, then passes that to this CallQueue, and AgentManager gets used by both. If time wasn't relevant I'd almost like to come back to this from the drawing board.

File: src/server.ts (hunk 25 above)
Comment (cejohnson-bw) [reply src/server.ts]: Yeah... the idea was to keep any V2 Voice stuff in here, but that encapsulation is already broken since we're getting BXML from the browser SDK inside of the AgentManager anyway. So, yeah, I would be fine with that.

File: src/webrtcAgentManager.ts (hunk 22 above)
Comment (cejohnson-bw) [reply src/webrtcAgentManager.ts]: So that's a neat little typescript feature I ran across that tells the compiler that the field is not null when it can't determine that itself. The `isOnline` function checks that, so rather than wrap this in another unnecessary `if (agent.conferenceId)` you can just pop a `!` on the end to tell typescript that you know what you're doing. Where that lands insofar as best practice I have no idea.

File: src/webrtcAgentManager.ts (hunk 23 above)
Comment (cejohnson-bw) [reply src/webrtcAgentManager.ts]: `getAgent` will throw an `AgentNotFoundError` if it can't find one. So you can rely on having an agent. Related, I don't have any error handling basically anywhere, which needs to be addressed.

File: src/webrtcAgentManager.ts (hunk 24 above)
Comment (cejohnson-bw) [reply src/webrtcAgentManager.ts]: I didn't call it something like that since it isn't actually doing the transfer; to actually perform the transfer it would have to know about V2 voice, which is even ickier than passing back BXML from the SDK I think. The caller id hack is really making it tough to encapsulate some of this.

File: src/webrtcAgentManager.ts (hunk 22 above)
Comment (nguyer) [reply src/webrtcAgentManager.ts]: TIL

File: src/VoiceCallManager.ts [hunk 26]
```diff
@@ -0,0 +1,178 @@
...
+          password: password
+        }
+      }
+    );
+  }
+
+  /**
+   * Connect a call into a WebRTC conference
+   * This method will make a POST to the HTTP Voice API to "redirect" the call,
+   * This will trigger another request for BXML to handle connecting the call
```
Comment (cejohnson-bw) [inline src/VoiceCallManager.ts]: Recommend replacing "This" with "which"

File: src/VoiceCallManager.ts [hunk 27]
```diff
@@ -0,0 +1,178 @@
...
+     * that sets the call's redirectUrl
+     */
+    this.router.post("/connect", async (req, res) => {
+      const callId = req.body.callId;
+      console.log(`attempting to connect call ${callId}`);
+      const tag = req.body.tag as string;
+
+      // This is a shortcut, so that we don't have to go look up the conference ID and participant ID again
+      const conferenceId = tag.split(",")[0];
+      const participantId = tag.split(",")[1];
```
Comment (cejohnson-bw) [inline src/VoiceCallManager.ts]: Interesting, didn't know about that feature.
